### PR TITLE
Add Types.createJsonQualifierImplementation

### PR DESCRIPTION
**Pull request by NightlyNexus**
From https://github.com/square/moshi/pull/236.
I don't know what this method should be called yet.

This is way too easy, so I am misunderstanding something. Can I get a nudge in the right direction?

### CODE REVIEW

File: moshi/src/main/java/com/squareup/moshi/Types.java [hunk 1]
```diff
@@ -137,6 +141,38 @@ static Type canonicalize(Type type) {
     }
   }
 
+  public static <T extends Annotation> T createJsonQualifierImplementation(
+      final Class<T> annotationType) {
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/Types.java]: Confirm that it’s an Annotation? `annotationType.isAnnotation()`

File: moshi/src/main/java/com/squareup/moshi/Types.java [hunk 2]
```diff
@@ -137,6 +141,38 @@ static Type canonicalize(Type type) {
...
+        new Class<?>[] { annotationType }, new InvocationHandler() {
+          @Override public Object invoke(Object proxy, Method method, Object[] args)
+              throws Throwable {
+            String methodName = method.getName();
+            if (methodName.equals("annotationType")) {
+              return annotationType;
+            }
+            if (methodName.equals("equals")) {
+              Object o = args[0];
+              return o.getClass().isInstance(annotationType);
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/Types.java]: dig

Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/Types.java]: oh, by the way `o` can be null!

File: moshi/src/main/java/com/squareup/moshi/Types.java [hunk 3]
```diff
@@ -137,6 +141,38 @@ static Type canonicalize(Type type) {
...
+            }
+            if (methodName.equals("hashCode")) {
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/Types.java]: dig

File: moshi/src/main/java/com/squareup/moshi/Types.java [hunk 4]
```diff
@@ -137,6 +141,38 @@ static Type canonicalize(Type type) {
...
+              return 0;
+            }
+            if (methodName.equals("toString")) {
+              return "@" + annotationType.getName() + "()";
+            }
+            throw new AssertionError();
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/Types.java]: I think you probably want just `return method.invoke(this, args)`. Or at least that’s what I’ve done elsewhere.

**Comment by swankjesse** [comment]
Yep, this is pretty much exactly what I want. But never satisfied, there’s a few more things I want.

The way I’m imagining this being exposed is not as a factory method for an annotation; but instead as an overload of `Moshi.adapter()`.

```
JsonAdapter<String> uppercaseAdapter = moshi.adapter(String.class, Uppercase.class);
```

The generated annotation instance will be used to resolve the type adapter.

File: moshi/src/main/java/com/squareup/moshi/Types.java [hunk 5]
```diff
@@ -137,6 +141,38 @@ static Type canonicalize(Type type) {
...
+    if (annotationType.getDeclaredMethods().length != 0) {
+      throw new IllegalArgumentException(annotationType + " must not declare methods.");
+    }
+    // noinspection unchecked
+    return (T) Proxy.newProxyInstance(annotationType.getClassLoader(),
+        new Class<?>[] { annotationType }, new InvocationHandler() {
+          @Override public Object invoke(Object proxy, Method method, Object[] args)
+              throws Throwable {
+            String methodName = method.getName();
+            if (methodName.equals("annotationType")) {
```
Comment (JakeWharton) [inline moshi/src/main/java/com/squareup/moshi/Types.java]: Use `switch`?

File: moshi/src/main/java/com/squareup/moshi/Types.java (hunk 4 above)
Comment (NightlyNexus) [inline moshi/src/main/java/com/squareup/moshi/Types.java]: Yep!
Why `this` instead of `proxy`?

Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/Types.java]: I think `proxy` would just stack overflow?

**Comment by NightlyNexus** [comment]
>as an overload of Moshi.adapter()

Cool, that's what I thought at first, but then I was unsure because:
- Should `nextAdapter` also have an overload?
- Should it be `adapter(Type, Set<Class<? extends Annotation>>)` or `adapter(Type, Class<? extends Annotation>)` The first won't be a valid overload (type erasure), and the second seems a little limiting or at least different from the current method.

It sounds like the answer is no and the latter, respectively, but I wanted to confirm.

**Comment by swankjesse** [comment]
I think the only one we need is `adapter(Type, Class<? extends Annotation>)`. I think it’s unlikely anyone will ever need multiple annotations for their entry point type.

**Comment by NightlyNexus** [comment]
Sounds good, I'll make tests, then!

**Comment by NightlyNexus** [comment]
Not really sure what a test on the `adapter` method should do, if necessary.

File: moshi/src/main/java/com/squareup/moshi/Moshi.java [hunk 6]
```diff
@@ -61,6 +61,11 @@
     return adapter(type, Util.NO_ANNOTATIONS);
   }
 
+  public <T> JsonAdapter<T> adapter(Type type, Class<? extends Annotation> annotationType) {
+    return adapter(type,
+        Collections.singleton(Types.createJsonQualifierImplementation(annotationType)));
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/Moshi.java]: nice

**Pull request merged**